tests/unittests/algo/test_evolution_es.py

```diff
@@ -0,0 +1,274 @@
+    space.register(Fidelity('epoch', 1, 8, 2))
+    return space
+
+
+@pytest.fixture
+def space2():
+    """Create a Space with two real dimensions."""
+    space = Space()
+    space.register(Real('lr', 'uniform', 0, 1))
+    space.register(Real('weight_decay', 'uniform', 0, 1))
+    return space
+
+
+@pytest.fixture
+def budgets():
+    """Return a configuration for a bracket."""
+    return [(30, 4), (30, 5), (30, 6)]
+
+
+@pytest.fixture
+def evolution(space1):
+    """Return an instance of EvolutionES."""
+    return EvolutionES(space1, repetitions=1, nums_population=4)
+
+
+@pytest.fixture
+def bracket(budgets, evolution, space1):
+    """Return a `Bracket` instance configured with `b_config`."""
+    return BracketEVES(evolution, budgets, 1, space1)
+
+
+@pytest.fixture
+def rung_0():
+    """Create fake points and objectives for rung 0."""
+    points = np.linspace(0, 8, 9)
+    return dict(
+        n_trials=9,
+        resources=1,
+        results={hashlib.md5(str([point]).encode('utf-8')).hexdigest(): (point, (1, point))
+                 for point in points})
+
+
+@pytest.fixture
+def rung_1(rung_0):
+    """Create fake points and objectives for rung 1."""
+    values = map(lambda v: (v[0], (3, v[0])), list(sorted(rung_0['results'].values()))[:3])
+    return dict(
+        n_trials=3,
+        resources=3,
+        results={hashlib.md5(str([value[0]]).encode('utf-8')).hexdigest(): value
+                 for value in values})
+
+
+@pytest.fixture
+def rung_2(rung_1):
+    """Create fake points and objectives for rung 2."""
+    values = map(lambda v: (v[0], (9, v[0])), list(sorted(rung_1['results'].values()))[:1])
+    return dict(
+        n_trials=1,
+        resources=9,
+        results={hashlib.md5(str([value[0]]).encode('utf-8')).hexdigest(): value
+                 for value in values})
+
+
+@pytest.fixture
+def rung_3():
+    """Create fake points and objectives for rung 3."""
+    points = np.linspace(1, 4, 4)
+    return dict(
+        n_trials=4,
+        resources=1,
+        results={hashlib.md5(str([point]).encode('utf-8')).hexdigest():
+                 (point, (np.power(2, (point - 1)), 1.0 / point, 1.0 / (point * point)))
+                 for point in points})
+
+
+@pytest.fixture
+def rung_4():
+    """Create duplicated fake points and objectives for rung 4."""
+    points = np.linspace(1, 4, 4)
+    return dict(
+        n_trials=4,
+        resources=1,
+        results={hashlib.md5(str([point]).encode('utf-8')).hexdigest():
+                 (point, (1, point // 2, point // 2))
+                 for point in points})
+
+
+def test_compute_budgets():
+    """Verify proper computation of budgets on a logarithmic scale"""
+    # Check typical values
+    assert compute_budgets(1, 3, 1, 2, 1) == [[(2, 1), (2, 2), (2, 3)]]
+    assert compute_budgets(1, 4, 2, 4, 2) == [[(4, 1), (4, 2), (4, 4)]]
+
+
+def test_customized_mutate_func(space2):
+    """Verify customized mutate function works correctly"""
```

Comment:
This test should be modified to use EvolutionES directly. If we modify the code in EvolutionES, in particular the code for the support of custom functions, we will need to synchronize any changes in the test here. This will likely lead to missed bugs. We should test the existing code in EvolutionES and avoid testing a duplication of its code otherwise maintainance and testing become hazardous.

Reply:
Good suggestion. Fixed.